lib/streamlit/elements/widgets/button.py

```diff
@@ -838,6 +854,16 @@ def _page_link(
                     main_script_directory=main_script_directory,
                 )
 
+        if query_params is not None:
+            if not hasattr(query_params, "get_all"):
+                query_params_internal: QueryParams = QueryParams()
+                # turn off sending forward messages for THIS QueryParams
+                query_params_internal._disable_forward_msg = True
+                query_params_internal.from_dict(query_params)
+            else:
+                query_params_internal = cast(QueryParams, query_params)
```

Comment:
As mentioned, do we need the `QueryParams` obj or could we generate the string directly?

Reply:
You _could_ generate the string correctly, however the `QueryParams` class has a number of protections and translation layers in place that make it safer, DRYer, and behaviorally consistent to just use the class to generate the string. For example, `QueryParams` does not allow the user to specify `embed` or any of the other embed-like keys, since those have special meaning. And `QueryParams` takes care of translating list-like objects in the incoming mapping into repeated keys in the query string.